pkg/app/controller.go

```diff
@@ -140,6 +153,9 @@ func newReverseProxyBuilder(im InstanceManager) ReverseProxyFactory {
 
 type HostForwarder struct {
 	newReverseProxy ReverseProxyFactory
+	getCredentials  func(UserInfo) (string, error)
```

Comment:
what do you think of coupling these two methods into a new abstraction, e.g RequestMutator and then we have InjectBuildAPICredentialsMutator which could be unit tested in isolation.

Reply:
Ended up going down a pretty deep rabbit hole addressing this comment. The gist is as follows:

- separating the mutator and host forwarder makes sense
- changing the original url for the host url is basically another kind of mutator
- with everything as a mutator the host forwarder became too simple: it just applied a list of mutators and called a reverse proxy
- so I deleted the (now too simple) host forwarder and rewrote the tests to interact directly with the controller instead
- the best way to test the controller is to give it the unix dabase and encryption services, but that introduced cyclical dependencies
- moved the interfaces and config types to its own package to eliminate the cycle